File: 2022/python/day06/solve.py
input = open("input.txt").read()


def first_marker(l):
    return next(filter(lambda i: len(set(input[i:i+l])) == l, range(len(input) - l + 1))) + l


print("Part 1:", first_marker(4))
print("Part 2:", first_marker(14))

File: 2022/haskell/day06/solve.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.Set (fromList)

firstMarker :: Int -> [Char] -> Int
firstMarker l input = head [i | i <- [0 ..], (length . fromList . take l . drop i $ input) == l] + l

solvePart1 :: String -> Int
solvePart1 = firstMarker 4

solvePart2 :: String -> Int
solvePart2 = firstMarker 14

main :: IO ()
main =
  do
    input <- readFile "input.txt"
    print input
    putStrLn $ "Part one: " ++ show (solvePart1 input)
    putStrLn $ "Part two: " ++ show (solvePart2 input)
